### DISCUSSION TITLE
WinUI 3 with C++ and XAML but without MIDL

### DISCUSSION THREAD

**Discussion by sotanakamura** [microsoft/microsoft-ui-xaml#8262] [Show and tell]
Hello everyone!

You can define WinUI 3 UIs with XAML. Classes reffered in XAML should be Windows runtime classes which you define with MIDL. Creating Windows Runtime classes is a little bit hard. I have written an article on how to define UIs with XAML but without MIDL.

* [WinUI 3 with C++ and XAML but without MIDL](https://github.com/sotanakamura/winui3-without-midl)

```xaml
// MainPage.xaml
<Window xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
    <StackPanel HorizontalAlignment="Center" VerticalAlignment="Center">
        <TextBlock>Hello World!</TextBlock>
    </StackPanel>
</Window>
```

```cpp
window = Window();
Application::LoadComponent(window, Uri(L"ms-appx:///MainPage.xaml"));
```

**Comment by driver1998** [comment]
If you really hate MIDL (or can't use it since you don't have VS or Windows SDK), you can make {Binding} work with an implementation of ICustomPropertyProvider.

This would seriously need some meta-programming or code-generation automation though, you won't want to write this by hand.

This is how I get binding working with MinGW (this is with XAML islands and System XAML though):

https://github.com/driver1998/XamlMinGW
https://github.com/driver1998/XamlMinGW/blob/26fab9825b77c0e1536fad0b050adce5a2f084f9/xaml-mingw/MainWindowViewModel.cpp#L57

**Reply by sotanakamura** [reply]
Nice! I will ckeck it!
